### PR TITLE
Dmd with control

### PR DESCRIPTION
**Pull request by eurika-kaiser**
Basic DMD with control implementation with toy example in tutorials
- New class DMDc
- Corresponding changes in Koopman class to account for additional control inputs
- state_transition_matrix corresponds to koopman_matrix
- control_matrix corresponds to control term or part of Koopman operator reflecting the control term
- The toy example may be used as test case, but not implemented

### CODE REVIEW

File: pykoopman/koopman.py [hunk 1]
```diff
@@ -157,3 +219,29 @@ def koopman_matrix(self):
         """
         check_is_fitted(self, "model")
         return self.model.steps[-1][1].coef_
+
+    @property
+    def state_transition_matrix(self):
+        """
+        The state transition matrix A satisfies x' = Ax + Bu.
+        # TODO: consider whether we want to match sklearn and have A and B satisfy
+        # x' = xA + uB instead
```
Comment (briandesilva) [inline pykoopman/koopman.py:228]: If we want to use the sklearn convention, we'll need to transpose our output.

File: pykoopman/regression/_dmdc.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+from ._base import BaseRegressor
+
+
+class DMDc(BaseRegressor):
+    """
+    DMD with control (DMDc) regressor.
+
+    Parameters
+    ----------
+    DMDc: BaseRegressor subclass
```
Comment (briandesilva) [inline pykoopman/regression/_dmdc.py:13]: I think we can remove this.
```suggestion
```

File: pykoopman/regression/_dmdc.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+from ._base import BaseRegressor
+
+
+class DMDc(BaseRegressor):
+    """
+    DMD with control (DMDc) regressor.
+
+    Parameters
+    ----------
```
Comment (briandesilva) [inline pykoopman/regression/_dmdc.py:12]: We'll want to fill out this docstring before merging in the pull request.

File: pykoopman/regression/_dmdc.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        -------
+        y: numpy ndarray, shape (n_samples, n_features)
+            Prediction of x one timestep in the future.
+
+        """
+        check_is_fitted(self, "coef_")
+        # TODO: this needs to be modified
+        # y = np.dot(self.state_matrix, x.T) + np.dot(self.control_matrix, u.T)
+        # return y.T
+        return self.regressor.predict(x.T).T
```
Comment (briandesilva) [inline pykoopman/regression/_dmdc.py:112]: This needs to be modified to use the control input.

```suggestion
        y = np.dot(self.state_matrix, x.T) + np.dot(self.control_matrix, u.T)
        return y.T
```

File: requirements-dev.txt [hunk 5]
```diff
@@ -16,3 +16,4 @@ sphinx_rtd_theme
 pre-commit
 sphinx-nbexamples
 jupyter_contrib_nbextensions
+osqp
```
Comment (briandesilva) [inline requirements-dev.txt:19]: Do we need this package? I don't see it being used anywhere.

**Comment by briandesilva** [comment]
I made some minor updates to the code. There are still a few outstanding issues we'll need to resolve before merging this in (see my comments in the files changed tab).

We'll also want to write some tests for DMDc before incorporating this into the larger package.

File: pykoopman/regression/_dmdc.py (hunk 4 above)
Comment (eurika-kaiser) [reply pykoopman/regression/_dmdc.py]: Fantastic, thanks a lot!

**Comment by eurika-kaiser** [comment]
Absolutely! I will continue working on that tomorrow and also include some tests.

**Pull request merged**